api/controllers/grades.go

```diff
@@ -6,283 +6,202 @@ import (
 	"time"
 
 	"github.com/UTDNebula/nebula-api/api/responses"
-
 	"github.com/gin-gonic/gin"
-
 	"go.mongodb.org/mongo-driver/bson"
 	"go.mongodb.org/mongo-driver/bson/primitive"
 	"go.mongodb.org/mongo-driver/mongo"
+	"go.mongodb.org/mongo-driver/mongo/options"
 )
 
-// We want to Filter (Match) ASAP
-
-// --------------------------------------------------------
-
-// Aggregate By Course -> Section:
-// ---- Prefix
-// ---- Prefix, Number
-// ---- Prefix, Number, SectionNumber
-
-// Aggregate By Professor -> Section
-// ---- Professor
+// GradesAggregation handles the aggregation of grades based on various filters.
+func GradesAggregation(c *gin.Context) {
+	var grades []map[string]interface{}
+	var results []map[string]interface{}
 
-// Aggregate By Find Course, Find Professor: then Match Section
-// ---- Prefix, Professor
-// ---- Prefix, Number, Professor
-// ---- Prefix, Number, Professor, SectionNumber
+	var cursor *mongo.Cursor
+	var collection *mongo.Collection
+	var pipeline mongo.Pipeline
 
-// --------------------------------------------------------
+	var err error
 
-// Filter on Course
-// ---- Prefix
-// ---- Prefix, Number
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
 
-// Filter on Course then Section
-// ---- Prefix, Number, SectionNumber
+	// Extract query parameters
+	prefix := c.Query("prefix")
+	number := c.Query("number")
+	sectionNumber := c.Query("section_number")
+	firstName := c.Query("first_name")
+	lastName := c.Query("last_name")
 
-// Filter on Professor
-// ---- Professor
+	professor := (firstName != "" || lastName != "")
 
-// Filter on Section by Matching Course and Professor IDs
-// ---- Prefix, Professor
-// ---- Prefix, Number, Professor
-// ---- Prefix, Number, Professor, SectionNumber
-
-// 4 Functions
+	// MongoDB client
+	client, err := mongo.Connect(ctx, options.Client().ApplyURI("mongodb://localhost:27017"))
+	if err != nil {
+		c.JSON(http.StatusInternalServerError, responses.GradeResponse{Status: http.StatusInternalServerError, Message: "error", Data: err.Error()})
+		return
+	}
+	defer func() {
+		if err = client.Disconnect(ctx); err != nil {
+			panic(err)
+		}
+	}()
+
+	// Database and collection
+	database := client.Database("your_database")
+	courseCollection := database.Collection("courses")
+	sectionCollection := database.Collection("sections")
+	professorCollection := database.Collection("professors")
+
+	// Stages for MongoDB aggregation pipeline
```

Comment:
Why are we connecting to mongo (again) inside of the controller? The collections are already connected elsewhere.

Doing this would also set up a fresh mongo connection for every request, which is inefficient.